### PR TITLE
ABL 52 issue

### PR DESCRIPTION
**Pull request by frostornge**
Sorry for late

* Custom **Time** system
* New user service
    * GetData (by UserId)
    * GetDataIds (by UserId)

### CODE REVIEW

File: collections/collection.go [hunk 1]
```diff
@@ -16,10 +14,10 @@ type Collection struct {
 
 	// format of the data the app want to collect
 	Schema    schemas.Schema
-	CreatedAt time.Time
+	CreatedAt common.Time
 }
 
-// IncentivePolicy is a policy about sharing revenue of the data
+// IcentivePolicy is a policy about sharing revenue of the data
```
Comment (therne) [inline collections/collection.go]: ?

File: proto/rpc/v1/server/warehouse.proto [hunk 2]
```diff
@@ -48,12 +51,15 @@ message EncryptedDataRequest {
     // ID of the user (Data Owner)
     string userId = 2;
 
+    // Timestamp of collected date
+    int64 collectedAt = 3;
```
Comment (therne) [inline proto/rpc/v1/server/warehouse.proto]: Why don't we use `uint` 😉

File: user/user.go [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+		}
+
+		for j, dataId := range info.DataIds {
+			bundle, err := manager.warehouse.Get(&dataId.DataId)
+			if err != nil {
+				return nil, errors.Wrap(err, "fetching bundle")
+			}
+
+			if _, ok := bundle.Data[dataId.UserId]; !ok {
+				return nil, fmt.Errorf("cannot find user data at given userId")
```
Comment (therne) [inline user/user.go]:
```suggestion
				return nil, errors.Errorf("cannot find user %s on given bundle", dataId.UserId)
```

File: user/user.go [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+			}
+
+			if uint32(len(bundle.Data[dataId.UserId])) <= dataId.RowId.Uint32() {
+				return nil, fmt.Errorf("cannot find row data at given userId")
+			}
+
+			encryptedData := bundle.Data[dataId.UserId][dataId.RowId.Uint32()]
+			if encryptedData.RowId.Uint32() != dataId.RowId.Uint32() {
+				return nil, fmt.Errorf(
```
Comment (therne) [inline user/user.go]:
```suggestion
				return nil, errors.Errorf(
```

File: user/user.go [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+			// dataIds
+			collection.DataIds = make([]struct {
+				common.DataId
+				CollectedAt int64 `json:"collectedAt"`
+			}, len(collection.RawDataIds))
+
+			for i, idPack := range collection.RawDataIds {
+				var rawDataId common.RawDataId
+				if err := mapstructure.Decode(idPack.Map(), &rawDataId); err != nil {
+					return nil, errors.Wrap(err, "decoding dataId")
```
Comment (therne) [inline user/user.go]:
```suggestion
					return nil, errors.Wrap(err, "failed to decode mongo document")
```

File: user/user.go [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+				}
+
+				dataId, err := rawDataId.Convert()
+				if err != nil {
+					return nil, errors.Wrap(err, "converting dataId")
```
Comment (therne) [inline user/user.go]:
```suggestion
					return nil, errors.Wrap(err, "failed to unmarshal data ID")
```

File: collections/collection.go (hunk 1 above)
Comment (frostornge) [reply collections/collection.go]: ?????

File: proto/rpc/v1/server/warehouse.proto (hunk 2 above)
Comment (frostornge) [reply proto/rpc/v1/server/warehouse.proto]: time.Now().UnixNano() returns int64

File: user/user.go (hunk 5 above)
Comment (frostornge) [reply user/user.go]: sry :)

File: user/user.go (hunk 6 above)
Comment (frostornge) [reply user/user.go]: GL

File: proto/rpc/v1/server/warehouse.proto (hunk 2 above)
Comment (therne) [reply proto/rpc/v1/server/warehouse.proto]: dg

Comment (therne) [reply proto/rpc/v1/server/warehouse.proto]: dz

File: collections/collection.go (hunk 1 above)
Comment (frostornge) [reply collections/collection.go]: patched

**Pull request merged**